main.go

```diff
@@ -1,17 +1,187 @@
+// Copyright 2024 Akamai Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
 package main
 
 import (
+	"context"
 	"flag"
+	"fmt"
+	"os"
+	"strconv"
+	"time"
 
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
 	"k8s.io/klog/v2"
+
+	metadata "github.com/linode/go-metadata"
+	decorator "github.com/linode/k8s-node-decorator/k8snodedecorator"
 )
 
-var version string
+var (
+	version  string
+	nodeName string
+)
 
 func init() {
 	_ = flag.Set("logtostderr", "true")
 }
 
+func GetCurrentNode(clientset *kubernetes.Clientset) (*corev1.Node, error) {
+	return clientset.CoreV1().Nodes().Get(context.TODO(), nodeName, metav1.GetOptions{})
+}
+
+func SetLabel(node *corev1.Node, key, newValue string) (changed bool) {
+	changed = false
+	oldValue, ok := node.Labels[key]
+
+	if !ok || oldValue != newValue {
+		changed = true
+		node.Labels[key] = newValue
+	}
+
+	return changed
+}
+
+func UpdateNodeLabels(
+	clientset *kubernetes.Clientset,
+	instanceData *metadata.InstanceData,
+) error {
+	if instanceData == nil {
+		return fmt.Errorf("instance data received from Linode metadata service is nil")
+	}
+
+	node, err := GetCurrentNode(clientset)
+	if err != nil {
+		return fmt.Errorf("failed to get the node: %w", err)
+	}
+
+	klog.Infof("Updating node labels with Linode instance data: %v", instanceData)
+	labelsUpdated := false
+
+	handleUpdated := func(updated bool) {
+		if updated {
+			labelsUpdated = updated
+		}
+	}
+
+	handleUpdated(SetLabel(node, "decorator.linode.com/label", instanceData.Label))
+	handleUpdated(SetLabel(node, "decorator.linode.com/instance-id", strconv.Itoa(instanceData.ID)))
+	handleUpdated(SetLabel(node, "decorator.linode.com/region", instanceData.Region))
+	handleUpdated(SetLabel(node, "decorator.linode.com/instance-type", instanceData.Type))
+	handleUpdated(SetLabel(node, "decorator.linode.com/host", instanceData.HostUUID))
+
+	for key, value := range decorator.ParseTags(instanceData.Tags) {
+		handleUpdated(SetLabel(node, key, value))
+	}
```

Comment:
I might be misreading this but it looks like this snippet might not drop deleted tags.

Maybe something like this would work? (untested)
```suggestion
	oldTags := make(map[string]string)
	maps.Copy(node.Labels, oldTags)
	
	newTags := decorator.ParseTags(instanceData.Tags)

	for key, value := range oldTags {
		if !strings.HasPrefix(key, decorator.TagLabelPrefix) {
			continue
		}

		// If the tag is no longer defined, drop it
		if _, ok := newTags[key]; !ok {
			delete(newTags, key)
			labelsUpdated = true
			continue
		}

		// Otherwise attempt to update the tag as normal
		handleUpdated(SetLabel(node, key, value))
	}
```

Reply:
Oops, thanks for discovering that! Will fix it

Reply:
The suggested change here doesn't seem to be adding new tags? Let me try to fix it based on this

Reply:
Oops you're right, my bad lol